### ISSUE TITLE
Update the best64 rules to be the same as hashcat's

### ISSUE THREAD

**Issue by solardiz** [magnumripper/JohnTheRipper#3202]
See this tweet thread: https://twitter.com/m33x/status/974345388701421568

Looks like we need to, and we can, simply take the latest best64 (78 rules?) from hashcat and use that in JtR as-is, now that we support the hashcat rule commands. Looks like (per testing done in that tweet thread) we don't even have to enable the hashcat rules mode in JtR for that (no incompatible commands are used), but more thorough testing (on more input words) is needed.

**Comment by frank-dittrich** [comment]
What if a user started a session using some older best64 rules, and restores with the updated?
Would it help to compute and store some hash of all the rules in a section, and warn when that hash changes on --restore? (This might require a new .rec file version.)
OTOH, unless the user kills the restored session immediately (i.e., before any changes are written to the status file), the "damage" is already been done, so the warning doesn't help too much.
Similar restore problems could occur if the word list file changes, etc.

**Comment by magnumripper** [comment]
I think we can ignore such cases. This is a password cracker, not an AED machine.

**Comment by magnumripper** [comment]
Using our default wordlist, adding `!! hashcat logic ON` doesn't change the output. I see that hashcat produces more words from that though, even after dropping the `#!comment:` lines and running through `uniq`. Not yet sure why.

**Comment by solardiz** [comment]
BTW, we could optimize these rules to produce fewer dupes by adding some `q` at the end (e.g., `u q`), `<*` at the start (e.g., `<* $1`), etc. where appropriate - like we have for our original rule sets. Maybe those changes would also be accepted by hashcat (@Chick3nman?), or if not we could maintain the hashcat-original rules under a separate name like best64-unoptimized for testing. OTOH, such dupe-avoidance will have performance impact when running against fast unsalted hashes.

**Comment by magnumripper** [comment]
You mean `Q` - or is there some lower-case `q` thing I'm not aware of?

**Comment by solardiz** [comment]
Oh, yes, I meant `Q`.

**Comment by magnumripper** [comment]
See #3208 for a bug that surfaced now.

**Comment by magnumripper** [comment]
After the fix for #3208, our best64.conf is a verbatim copy of hashcat's best64.rule.

**Comment by magnumripper** [comment]
Here's what I tested. Before I added the `sort -u` I realized that hashcat runs all rules over first word before going on to next word. JtR does it the other way round: It runs all words through first rule before going on to next rule.
```
$ cd hashcat
$ grep -vF '#!comment:' ../john/run/password.lst | ./hashcat -a0 -r rules/best64.rule --stdout | LC_ALL=C sort -u > hc_b64.txt
$ cd ../john/run
$ ./john -w -ru:best64 -stdout | LC_ALL=C sort -u >jtr_b64.txt
$ diff jtr_b64.txt ../../hashcat/hc_b64.txt
11256a11257
> 8
```
So one line of difference: Hashcat ended up with just the string "8" from some word w/ some rule, while JtR did not. Using hashcat's "--debug-mode=3" showed that the original word was the empty string and the rule was "$8" which makes sense. Hashcat actually produces 35 unique words from the empty string with best64 rules, but all the other ones were dupes in this case.

The difference is due to the fact JtR doesn't apply rules to an empty word. I verified that "john proper" also doesn't do that.

So I changed our logic: We now do apply rules to empty words if "hashcat logic" is turned on. After this, both tools produce the exact same (net) output.
